core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -86,14 +89,69 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
+   * @throws ControllerMovedException if fail to create /controller or fail to increment controller epoch.
+   */
+  def registerControllerAndIncrementControllerEpoch(controllerId: Int): (Int, Int) = {
+    val timestamp = time.milliseconds()
+
+    // Read /controller_epoch to get the current controller epoch and zkVersion,
+    // create /controller_epoch with initial value if not exists
+    val (curEpoch, curEpochZkVersion) = getControllerEpoch
+      .map(e => (e._1, e._2.getVersion))
+      .getOrElse(createControllerEpochZNode())
+
+    // Create /controller and update /controller_epoch atomically
+    val newControllerEpoch = curEpoch + 1
+    val expectedControllerEpochZkVersion = curEpochZkVersion
+
+    debug(s"Try to create ${ControllerZNode.path} and increment controller epoch to $newControllerEpoch with expected controller epoch zkVersion $expectedControllerEpochZkVersion")
+    def tryCreateControllerZNodeAndIncrementEpoch(): (Int, Int) = {
+      try {
+        val transaction = zooKeeperClient.createTransaction()
+        transaction.create(ControllerZNode.path, ControllerZNode.encode(controllerId, timestamp),
+          acls(ControllerZNode.path).asJava, CreateMode.EPHEMERAL)
+        transaction.setData(ControllerEpochZNode.path, ControllerEpochZNode.encode(newControllerEpoch), expectedControllerEpochZkVersion)
+        val results = transaction.commit()
+        val setDataResult = results.get(1).asInstanceOf[SetDataResult]
+        (newControllerEpoch, setDataResult.getStat.getVersion)
+      } catch {
+        case _: NodeExistsException =>
+          val curControllerId = getControllerId.getOrElse(throw new ControllerMovedException(
+            s"The ephemeral node at ${ControllerZNode.path} went away while checking whether the controller election succeeds. " +
+              s"Aborting controller startup procedure"))
+          if (controllerId == curControllerId) {
+            val (epoch, stat)  = getControllerEpoch.getOrElse(throw new ControllerMovedException(
+              "Controller moved to another broker. Aborting controller startup procedure"))
+            // If the epoch is the same as newControllerEpoch, it is safe to infer that the returned epoch zkVersion
+            // is associated with the current broker during controller election because we already knew that the zk
+            // transaction succeeds based on the controller znode verification. Other rounds of controller
+            // election will result in larger epoch number written in zk.
+            if (epoch == newControllerEpoch)
+              return (newControllerEpoch, stat.getVersion)
+          }
+          throw new ControllerMovedException("Controller moved to another broker. Aborting controller startup procedure")
+        case _: BadVersionException =>
+          throw new ControllerMovedException("Controller moved to another broker. Aborting controller startup procedure")
```

Comment:
According to ZK doc, multi propagates the error from one of the operations. The BadVersionException could be the result of a retry after ConnectionLossException. So, it seems that we need to handle it in the same way as NodeExistsException.

Reply:
Correct me if I am wrong, I think there are two cases when we see BadVersionException here:
1. Another round of controller election kicks in and the controller does switch. It is safe to throw `ControllerMovedException` in this case.
2. The current broker loss zk connection after zk successfully finishes the transaction, **and the controller znode is gone**. In this case, another round of controller election will be triggered by zk watcher `handleDeleted`. So I think it is also safe to throw `ControllerMovedException` here.

Reply:
Thanks for the prompt reply!
I actually checked both client and server side codes of zookeeper, the implementation honors the order and the thrown exception will correspond to the first error it sees. But since the muli api doesn't explicitly say that it is the case or it will maintain this guarantee in the future, I agree to also handle BadVersion the same way as NodeExists for safety given that the performance overhead is little.

Reply:
Done.